File: sdk/src/iosMain/kotlin/com/streamprobe/sdk/internal/AVPlayerProbe.kt
package com.streamprobe.sdk.internal

import com.streamprobe.sdk.model.ActiveTrackInfo
import com.streamprobe.sdk.model.SwitchReason
import com.streamprobe.sdk.model.TrackSwitchEvent
import com.streamprobe.sdk.model.TracksSnapshot
import kotlinx.cinterop.ExperimentalForeignApi
import platform.AVFoundation.AVAssetVariant
import platform.AVFoundation.AVKeyValueStatusLoaded
import platform.AVFoundation.AVMediaCharacteristicAudible
import platform.AVFoundation.AVMediaCharacteristicLegible
import platform.AVFoundation.AVMediaSelectionOption
import platform.AVFoundation.AVPlayer
import platform.AVFoundation.AVPlayerItemAccessLogEvent
import platform.AVFoundation.AVPlayerItemDidPlayToEndTimeNotification
import platform.AVFoundation.AVPlayerItemErrorLogEvent
import platform.AVFoundation.AVPlayerItemNewAccessLogEntryNotification
import platform.AVFoundation.AVPlayerItemNewErrorLogEntryNotification
import platform.AVFoundation.AVURLAsset
import platform.AVFoundation.accessLog
import platform.AVFoundation.asset
import platform.AVFoundation.currentItem
import platform.AVFoundation.currentMediaSelection
import platform.AVFoundation.errorLog
import platform.AVFoundation.mediaSelectionGroupForMediaCharacteristic
import platform.AVFoundation.variants
import platform.Foundation.NSDate
import platform.Foundation.NSNotificationCenter
import platform.Foundation.NSOperationQueue
import platform.Foundation.timeIntervalSince1970
import platform.QuartzCore.CACurrentMediaTime
import platform.darwin.NSObjectProtocol
import platform.darwin.dispatch_async
import platform.darwin.dispatch_get_main_queue
import kotlin.math.abs

/**
 * iOS analogue of `PlayerInterceptor`: observes an `AVPlayer` and writes the existing SDK models
 * into the shared [SessionStore]. Headless — no UI.
 *
 * - Track listing (3.3): async `variants`-key load → `updateTrackList` (parsing the HLS master is
 *   enough, so it works before `readyToPlay` and without playback progressing — there is no
 *   KVO/time-observer because playback does not progress when headless). Stale-closure guard: the
 *   load callback only publishes when `statusOfValueForKey(...) == Loaded` **and** the live
 *   `this.player?.currentItem?.asset === asset` (the live player, not a captured local), so a stale
 *   or re-attached closure is rejected.
 * - Segment metrics / bitrate switches / dropped frames (3.4–3.6): each new `AVPlayerItemAccessLog`
 *   entry is mapped to a coarse `SegmentMetric`, an adaptive `VideoSwitch` (on `indicatedBitrate`
 *   change across entries), and a `DROPPED_FRAMES` error. The *current* (still-mutating) log entry is
 *   processed one notification later (`accessLog()?.events?.dropLast(1)`), once it is finalized, so
 *   only complete (non-mutating) stats are read; the last entry is picked up by the next notification.
 * - Errors (3.6): each new `AVPlayerItemErrorLog` entry → `LOAD_ERROR`.
 * - Active track resolution: `emitBitrateSwitch` looks up the closest variant in the known list to
 *   provide resolution/codecs in `ActiveTrackInfo` (access log alone only carries bitrate). It also
 *   marks `isSelected` on the matching `VariantInfo` in the stored snapshot.
 * - Audio selection: `publishTracks` reads `currentMediaSelection` to detect the initially-selected
 *   audio rendition. `refreshAudioSelection` is called from `onNewAccessLogEntry` as a fallback for
 *   streams where the selection is made after the variants key loads.
 *
 * Observers are registered via `NSNotificationCenter` scoped to [AVPlayer.currentItem] (the object
 * filter) so sibling-player / `AVQueuePlayer` callbacks cannot fire this probe's handlers. All
 * callbacks are delivered on the main queue. [nowMs] uses `CACurrentMediaTime()` plus an epoch offset
 * computed once at [attach] for monotonic timestamps (avoids wall-clock non-monotonicity).
 * [attach] calls `store.clear()`; the [NSObjectProtocol] observer tokens are removed in [detach].
 *
 * **Known iOS limitation:** Observers and [finalizedAccessEntries] / [processedErrorEntries] counters
 * are scoped to the [AVPlayer.currentItem] at [attach] time. `AVQueuePlayer` item changes are not
 * detected — metrics from subsequent items continue to accumulate but the counters may skip entries.
 * Full `AVQueuePlayer` support requires re-attaching on `AVPlayerItem` replacement (Phase 4+).
 */
internal class AVPlayerProbe(
    private val store: SessionStore,
) {
    private var player: AVPlayer? = null
    private var active = false
    private val observers = mutableListOf<NSObjectProtocol>()

    private var finalizedAccessEntries = 0
    private var processedErrorEntries = 0
    private var lastIndicatedBitrate = -1.0

    // Monotonic clock anchor: computed once per attach() so nowMs() is wall-clock but never goes backward.
    private var epochOffsetMs: Long = 0L

    @OptIn(ExperimentalForeignApi::class)
    fun attach(player: AVPlayer) {
        detach()
        this.player = player
        active = true
        epochOffsetMs = (NSDate().timeIntervalSince1970 * MILLIS_PER_SECOND).toLong() -
            (CACurrentMediaTime() * MILLIS_PER_SECOND).toLong()
        registerLogObservers()
        val asset = player.currentItem?.asset as? AVURLAsset ?: return
        asset.loadValuesAsynchronouslyForKeys(listOf(VARIANTS_KEY)) {
            dispatch_async(dispatch_get_main_queue()) {
                // Guard: probe still active AND this is still the current asset (item may have changed
                // or a re-attach may have installed a different player entirely).
                if (active && this.player?.currentItem?.asset == asset) {
                    val status = asset.statusOfValueForKey(VARIANTS_KEY, null)
                    if (status == AVKeyValueStatusLoaded) publishTracks(asset)
                }
            }
        }
    }

    fun detach() {
        // Flush any unprocessed entries before tearing down (covers the manual-stop case; the
        // natural-end case is already handled by the AVPlayerItemDidPlayToEndTimeNotification observer).
        onNewAccessLogEntry(flushAll = true)
        active = false
        val center = NSNotificationCenter.defaultCenter
        observers.forEach { center.removeObserver(it) }
        observers.clear()
        player = null
        finalizedAccessEntries = 0
        processedErrorEntries = 0
        lastIndicatedBitrate = -1.0
        epochOffsetMs = 0L
    }

    private fun registerLogObservers() {
        val center = NSNotificationCenter.defaultCenter
        val main = NSOperationQueue.mainQueue
        // Scope observers to the current item so multi-player or AVQueuePlayer siblings don't
        // trigger this probe's callbacks. If currentItem is null, falls back to null (fires for all).
        val currentItem = player?.currentItem
        observers +=
            center.addObserverForName(AVPlayerItemNewAccessLogEntryNotification, currentItem, main) { _ ->
                // AVPlayerItemNewAccessLogEntryNotification fires when a new entry is STARTED, meaning
                // the last entry in events[] is still accumulating. Only process entries before it.
                if (active) onNewAccessLogEntry(flushAll = false)
            }
        observers +=
            center.addObserverForName(AVPlayerItemNewErrorLogEntryNotification, currentItem, main) { _ ->
                if (active) onNewErrorLogEntry()
            }
        observers +=
            center.addObserverForName(AVPlayerItemDidPlayToEndTimeNotification, currentItem, main) { _ ->
                // Playback ended naturally — the last access-log entry is now final; flush it.
                if (active) onNewAccessLogEntry(flushAll = true)
            }
    }

    private fun publishTracks(asset: AVURLAsset) {
        val variants = asset.variants.mapNotNull { (it as? AVAssetVariant)?.let(::mapVariant) }
        if (variants.isEmpty()) return
        val mediaSelection = player?.currentItem?.currentMediaSelection
        val audioGroup = asset.mediaSelectionGroupForMediaCharacteristic(AVMediaCharacteristicAudible)
        val selectedAudio = audioGroup?.let { mediaSelection?.selectedMediaOptionInMediaSelectionGroup(it) }
        val audioTracks =
            audioGroup
                ?.options
                ?.filterIsInstance<AVMediaSelectionOption>()
                .orEmpty()
                .map { option -> mapAudioOption(option).copy(isSelected = option == selectedAudio) }
        val subtitleGroup = asset.mediaSelectionGroupForMediaCharacteristic(AVMediaCharacteristicLegible)
        val selectedSubtitle = subtitleGroup?.let { mediaSelection?.selectedMediaOptionInMediaSelectionGroup(it) }
        val subtitleTracks =
            subtitleGroup
                ?.options
                ?.filterIsInstance<AVMediaSelectionOption>()
                .orEmpty()
                .map { option -> mapLegibleOption(option).copy(isSelected = option == selectedSubtitle) }
        store.updateTrackList(TracksSnapshot(variants, audioTracks, subtitleTracks))
        store.updateActiveAudioTrack(audioTracks.find { it.isSelected })
        store.updateActiveSubtitleTrack(subtitleTracks.find { it.isSelected })
    }

    private fun onNewAccessLogEntry(flushAll: Boolean = false) {
        val events = player?.currentItem?.accessLog()?.events ?: return
        // When flushAll=false (new-entry notification): the just-added last entry is still
        // accumulating, so only entries before it are final. When flushAll=true (end-of-playback
        // or detach): no further entries will arrive, so the last entry is also final.
        val finalized = if (flushAll) events.size else events.size - 1
        while (finalizedAccessEntries < finalized) {
            (events[finalizedAccessEntries] as? AVPlayerItemAccessLogEvent)?.let(::processAccessEntry)
            finalizedAccessEntries++
        }
        // Access log firing means the player is actively playing — a reliable point to detect the
        // selected audio rendition if publishTracks fired before the player made its initial selection.
        if (store.activeAudioTrack.value == null) refreshAudioSelection()
    }

    private fun processAccessEntry(event: AVPlayerItemAccessLogEvent) {
        // Use the entry's own start date as requestTimestampMs — equivalent to Android's
        // System.currentTimeMillis() - loadEventInfo.loadDurationMs. Falls back to nowMs() when
        // playbackStartDate is nil (which AVFoundation documents as possible for the first entry).
        val requestTimestampMs =
            event.playbackStartDate
                ?.let { (it.timeIntervalSince1970 * MILLIS_PER_SECOND).toLong() }
                ?: nowMs()
        store.addSegmentMetric(
            accessLogSegmentMetric(
                nowMs = requestTimestampMs,
                uri = event.URI,
                sizeBytes = event.numberOfBytesTransferred,
                observedBitrate = event.observedBitrate,
                transferDurationSeconds = event.transferDuration,
            ),
        )
        emitBitrateSwitch(requestTimestampMs, event.indicatedBitrate)
        if (event.numberOfDroppedVideoFrames >= DROPPED_FRAME_THRESHOLD) {
            store.addPlaybackError(droppedFramesError(requestTimestampMs, event.numberOfDroppedVideoFrames))
        }
    }

    private fun emitBitrateSwitch(
        nowMs: Long,
        indicatedBitrate: Double,
    ) {
        if (!isBitrateSwitch(lastIndicatedBitrate, indicatedBitrate)) return
        val newTrack = resolveActiveTrack(indicatedBitrate)
        val previousTrack = store.activeTrack.value
        val reason = if (previousTrack == null) SwitchReason.INITIAL else SwitchReason.ADAPTIVE
        store.addTrackSwitchEvent(
            TrackSwitchEvent.VideoSwitch(
                timestampMs = nowMs,
                bufferDurationMs = 0,
                reason = reason,
                previousTrack = previousTrack,
                newTrack = newTrack,
            ),
        )
        store.updateActiveTrack(newTrack)
        updateVariantSelection(newTrack.bitrate)
        lastIndicatedBitrate = indicatedBitrate
    }

    /**
     * Looks up the closest-bitrate variant in the known track list, providing resolution and codecs
     * that the access log's `indicatedBitrate` alone cannot supply. Falls back to a bitrate-only
     * [ActiveTrackInfo] when no track list has been published yet.
     */
    private fun resolveActiveTrack(indicatedBitrate: Double): ActiveTrackInfo {
        val bitrate = indicatedBitrate.toInt()
        val variant =
            store.trackListInfo.value
                ?.variants
                ?.minByOrNull { abs(it.bitrate - bitrate) }
        return if (variant != null) {
            ActiveTrackInfo(
                bitrate = variant.bitrate,
                width = variant.width,
                height = variant.height,
                codecs = variant.codecs,
                id = variant.id,
            )
        } else {
            activeTrackFromIndicatedBitrate(indicatedBitrate)
        }
    }

    /** Marks the variant closest in bitrate to [selectedBitrate] as selected in the stored snapshot. */
    private fun updateVariantSelection(selectedBitrate: Int) {
        val snapshot = store.trackListInfo.value ?: return
        val idx =
            snapshot.variants.indices
                .minByOrNull { i -> abs(snapshot.variants[i].bitrate - selectedBitrate) } ?: return
        val updated = snapshot.variants.mapIndexed { i, v -> v.copy(isSelected = i == idx) }
        store.updateTrackList(TracksSnapshot(updated, snapshot.audioTracks, snapshot.subtitleTracks))
    }

    /**
     * Queries [AVPlayerItem.currentMediaSelection] to detect which audio rendition is playing and
     * updates [SessionStore] with the result. Called lazily from [onNewAccessLogEntry] for streams
     * where the player makes its audio selection after the variants key has loaded.
     */
    private fun refreshAudioSelection() {
        val currentItem = player?.currentItem ?: return
        val asset = currentItem.asset as? AVURLAsset ?: return
        val snapshot = store.trackListInfo.value ?: return
        val group = asset.mediaSelectionGroupForMediaCharacteristic(AVMediaCharacteristicAudible) ?: return
        val selected = currentItem.currentMediaSelection.selectedMediaOptionInMediaSelectionGroup(group) ?: return
        val options = group.options.filterIsInstance<AVMediaSelectionOption>()
        val idx = options.indexOfFirst { it == selected }
        if (idx < 0 || snapshot.audioTracks.getOrNull(idx)?.isSelected == true) return
        val updatedAudio = snapshot.audioTracks.mapIndexed { i, t -> t.copy(isSelected = i == idx) }
        store.updateTrackList(TracksSnapshot(snapshot.variants, updatedAudio, snapshot.subtitleTracks))
        store.updateActiveAudioTrack(updatedAudio.find { it.isSelected })
    }

    private fun onNewErrorLogEntry() {
        val events = player?.currentItem?.errorLog()?.events ?: return
        while (processedErrorEntries < events.size) {
            (events[processedErrorEntries] as? AVPlayerItemErrorLogEvent)?.let { event ->
                store.addPlaybackError(
                    loadError(
                        nowMs = nowMs(),
                        errorDomain = event.errorDomain,
                        statusCode = event.errorStatusCode,
                        uri = event.URI,
                        comment = event.errorComment,
                    ),
                )
            }
            processedErrorEntries++
        }
    }

    // Monotonic wall-clock: CACurrentMediaTime() is mach_absolute_time-based and never goes backward.
    // epochOffsetMs bridges it to epoch time (set once per attach so NTP adjustments don't affect
    // relative timestamps within a session).
    private fun nowMs(): Long = (CACurrentMediaTime() * MILLIS_PER_SECOND).toLong() + epochOffsetMs

    private companion object {
        const val VARIANTS_KEY = "variants"
        const val DROPPED_FRAME_THRESHOLD = 3
    }
}

File: Sources/StreamProbe/AVPlayerProbe.swift
import AVFoundation
import QuartzCore
import StreamProbeCore

/// iOS analogue of the Kotlin `AVPlayerProbe`: observes an `AVPlayer` and writes the SDK models into
/// the Core through a write-only `DiagnosticsSink`. Headless — no UI.
///
/// Because the sink is write-only, the probe keeps its own mirror of the state the Kotlin version read
/// back from the store (`currentSnapshot`, `currentActiveTrack`, `hasSelectedAudio`). The probe is the
/// sole writer on iOS, so its local copy is authoritative.
///
/// Not thread-safe: all callbacks arrive on the main queue (notification observers use `.main`; the
/// variants-load completion hops to main). The iOS 18 AVMetrics consumer task runs off the main thread
/// between awaits, but all `sink` access remains pinned to `MainActor`. Holds the player weakly, but on
/// iOS 18+ the AVMetrics consumer `Task` retains the current `AVPlayerItem` until `detach()` cancels it,
/// so the host must `detach()` before releasing the player.
final class AVPlayerProbe {
    private let sink: DiagnosticsSink
    private weak var player: AVPlayer?
    private var active = false
    private var observers: [NSObjectProtocol] = []

    private var processedEventEntries = 0
    private var finalizedSegmentEntries = 0
    private var processedErrorEntries = 0
    private var lastIndicatedBitrate = -1.0

    private var currentSnapshot: TracksSnapshot?
    private var currentActiveTrack: ActiveTrackInfo?
    private var hasSelectedAudio = false

    private var epochOffsetMs: Int64 = 0
    private var metricsTask: Task<Void, Never>?

    private let variantsKey = "variants"
    private let droppedFrameThreshold = 3

    init(sink: DiagnosticsSink) {
        self.sink = sink
    }

    func attach(player: AVPlayer) {
        // Re-attaching the same player instance would duplicate observers and the metrics consumer; no-op.
        // (Item swaps on the same player are not re-observed — item-change tracking is not yet supported.)
        if active, self.player === player { return }
        detach()
        self.player = player
        active = true
        epochOffsetMs = Int64(Date().timeIntervalSince1970 * 1000) - Int64(CACurrentMediaTime() * 1000)
        registerLogObservers()
        startMetricsConsumer()
        guard let asset = player.currentItem?.asset as? AVURLAsset else { return }
        asset.loadValuesAsynchronously(forKeys: [variantsKey]) { [weak self] in
            DispatchQueue.main.async {
                guard let self, self.active, self.player?.currentItem?.asset === asset else { return }
                var error: NSError?
                if asset.statusOfValue(forKey: self.variantsKey, error: &error) == .loaded {
                    self.publishTracks(asset)
                }
            }
        }
    }

    func detach() {
        // Flush pending entries before teardown (manual-stop case; natural end is handled by the
        // DidPlayToEndTime observer).
        onNewAccessLogEntry()
        flushLastSegmentEntry()
        active = false
        let center = NotificationCenter.default
        observers.forEach { center.removeObserver($0) }
        observers.removeAll()
        metricsTask?.cancel()
        metricsTask = nil
        player = nil
        processedEventEntries = 0
        finalizedSegmentEntries = 0
        processedErrorEntries = 0
        lastIndicatedBitrate = -1.0
        currentSnapshot = nil
        currentActiveTrack = nil
        hasSelectedAudio = false
        epochOffsetMs = 0
    }

    private func registerLogObservers() {
        let center = NotificationCenter.default
        let item = player?.currentItem
        observers.append(center.addObserver(forName: .AVPlayerItemNewAccessLogEntry, object: item, queue: .main) { [weak self] _ in
            guard let self, self.active else { return }
            self.onNewAccessLogEntry()
        })
        observers.append(center.addObserver(forName: .AVPlayerItemNewErrorLogEntry, object: item, queue: .main) { [weak self] _ in
            guard let self, self.active else { return }
            self.onNewErrorLogEntry()
        })
        observers.append(center.addObserver(forName: .AVPlayerItemDidPlayToEndTime, object: item, queue: .main) { [weak self] _ in
            guard let self, self.active else { return }
            self.onNewAccessLogEntry()
            self.flushLastSegmentEntry()
        })
    }

    private func startMetricsConsumer() {
        guard #available(iOS 18.0, *), let item = player?.currentItem else { return }
        metricsTask = Task { [weak self] in
            do {
                for try await event in item.metrics(forType: AVMetricHLSMediaSegmentRequestEvent.self) {
                    await MainActor.run { [weak self] in
                        guard let self, self.active else { return }
                        self.sink.addSegmentMetric(metric: AVMetricsSegmentAdapter.segmentMetric(from: event, nowMs: self.nowMs()))
                    }
                }
            } catch is CancellationError {
                // Expected teardown path: detach() cancelled the task.
            } catch {
                // AVFoundation metrics stream error; swallowed so the debug probe never crashes the host app.
                // TODO(future release): surface/handle this instead of going silent — on iOS 18+ there is no
                // access-log fallback for segment metrics, so a dead stream means no per-segment data at all.
            }
        }
    }

    private func publishTracks(_ asset: AVURLAsset) {
        let variants = asset.variants.map { mapVariant($0) }
        guard !variants.isEmpty else { return }
        let mediaSelection = player?.currentItem?.currentMediaSelection
        let audioGroup = asset.mediaSelectionGroup(forMediaCharacteristic: .audible)
        let selectedAudio = audioGroup.flatMap { mediaSelection?.selectedMediaOption(in: $0) }
        let audioTracks = (audioGroup?.options ?? []).map { mapAudioOption($0, isSelected: $0.isEqual(selectedAudio)) }
        let subtitleGroup = asset.mediaSelectionGroup(forMediaCharacteristic: .legible)
        let selectedSubtitle = subtitleGroup.flatMap { mediaSelection?.selectedMediaOption(in: $0) }
        let subtitleTracks = (subtitleGroup?.options ?? []).map { mapLegibleOption($0, isSelected: $0.isEqual(selectedSubtitle)) }
        let snapshot = TracksSnapshot(variants: variants, audioTracks: audioTracks, subtitleTracks: subtitleTracks)
        sink.updateTrackList(info: snapshot)
        currentSnapshot = snapshot
        let activeAudio = audioTracks.first { $0.isSelected }
        sink.updateActiveAudioTrack(info: activeAudio)
        hasSelectedAudio = activeAudio != nil
        sink.updateActiveSubtitleTrack(info: subtitleTracks.first { $0.isSelected })
    }

    private func onNewAccessLogEntry() {
        guard let events = player?.currentItem?.accessLog()?.events else { return }
        // Bitrate switches and dropped frames: process every entry immediately, including the live tail.
        // These fields stabilise as soon as the entry is created, so waiting adds visible overlay lag.
        while processedEventEntries < events.count {
            let event = events[processedEventEntries]
            let tsMs = event.playbackStartDate.map { Int64($0.timeIntervalSince1970 * 1000) } ?? nowMs()
            emitBitrateSwitch(nowMs: tsMs, indicatedBitrate: event.indicatedBitrate)
            if event.numberOfDroppedVideoFrames >= droppedFrameThreshold {
                sink.addPlaybackError(event: AVAccessLogMappersKt.droppedFramesError(
                    nowMs: tsMs,
                    droppedFrames: Int64(event.numberOfDroppedVideoFrames)
                ))
            }
            processedEventEntries += 1
        }
        // Segment metrics (pre-iOS 18 only): defer the last entry because AVFoundation is still
        // accumulating bytes, transfer duration, and observed bitrate into it. Each new entry
        // supersedes the previous tail, making it safe to emit.
        if #unavailable(iOS 18.0) {
            while finalizedSegmentEntries < max(0, events.count - 1) {
                emitSegmentMetric(events[finalizedSegmentEntries])
                finalizedSegmentEntries += 1
            }
        }
        if !hasSelectedAudio { refreshAudioSelection() }
    }

    // Emits the deferred last segment-metric entry. Called on natural end-of-playback and on detach
    // so the final segment is never silently dropped.
    private func flushLastSegmentEntry() {
        guard #unavailable(iOS 18.0),
              let events = player?.currentItem?.accessLog()?.events,
              finalizedSegmentEntries < events.count else { return }
        emitSegmentMetric(events[finalizedSegmentEntries])
        finalizedSegmentEntries += 1
    }

    private func emitSegmentMetric(_ event: AVPlayerItemAccessLogEvent) {
        let tsMs = event.playbackStartDate.map { Int64($0.timeIntervalSince1970 * 1000) } ?? nowMs()
        sink.addSegmentMetric(metric: AVAccessLogMappersKt.accessLogSegmentMetric(
            nowMs: tsMs,
            uri: event.uri,
            sizeBytes: event.numberOfBytesTransferred,
            observedBitrate: event.observedBitrate,
            transferDurationSeconds: event.transferDuration
        ))
    }

    private func emitBitrateSwitch(nowMs: Int64, indicatedBitrate: Double) {
        guard AVAccessLogMappersKt.isBitrateSwitch(previousIndicated: lastIndicatedBitrate, currentIndicated: indicatedBitrate) else { return }
        let newTrack = resolveActiveTrack(indicatedBitrate)
        let previousTrack = currentActiveTrack
        let reason: SwitchReason = previousTrack == nil ? .initial : .adaptive
        sink.addTrackSwitchEvent(event: TrackSwitchEventVideoSwitch(
            timestampMs: nowMs,
            bufferDurationMs: 0,
            reason: reason,
            previousTrack: previousTrack,
            newTrack: newTrack
        ))
        sink.updateActiveTrack(info: newTrack)
        currentActiveTrack = newTrack
        updateVariantSelection(newTrack.bitrate)
        lastIndicatedBitrate = indicatedBitrate
    }

    private func resolveActiveTrack(_ indicatedBitrate: Double) -> ActiveTrackInfo {
        let target = Int32(indicatedBitrate)
        if let variant = currentSnapshot?.variants.min(by: { abs($0.bitrate - target) < abs($1.bitrate - target) }) {
            return ActiveTrackInfo(bitrate: variant.bitrate, width: variant.width, height: variant.height, codecs: variant.codecs, id: variant.id)
        }
        return AVAccessLogMappersKt.activeTrackFromIndicatedBitrate(indicatedBitrate: indicatedBitrate)
    }

    private func updateVariantSelection(_ selectedBitrate: Int32) {
        guard let snapshot = currentSnapshot, !snapshot.variants.isEmpty else { return }
        let variants = snapshot.variants
        var bestIdx = 0
        for i in variants.indices where abs(variants[i].bitrate - selectedBitrate) < abs(variants[bestIdx].bitrate - selectedBitrate) {
            bestIdx = i
        }
        let updated = variants.enumerated().map { index, v in
            VariantInfo(bitrate: v.bitrate, width: v.width, height: v.height, codecs: v.codecs, frameRate: v.frameRate, id: v.id, isSelected: index == bestIdx)
        }
        let newSnapshot = TracksSnapshot(variants: updated, audioTracks: snapshot.audioTracks, subtitleTracks: snapshot.subtitleTracks)
        sink.updateTrackList(info: newSnapshot)
        currentSnapshot = newSnapshot
    }

    private func refreshAudioSelection() {
        guard let currentItem = player?.currentItem,
              let asset = currentItem.asset as? AVURLAsset,
              let snapshot = currentSnapshot,
              let group = asset.mediaSelectionGroup(forMediaCharacteristic: .audible),
              let selected = currentItem.currentMediaSelection.selectedMediaOption(in: group),
              let idx = group.options.firstIndex(where: { $0.isEqual(selected) }) else { return }
        if snapshot.audioTracks.indices.contains(idx), snapshot.audioTracks[idx].isSelected { return }
        let updatedAudio = snapshot.audioTracks.enumerated().map { index, t in
            AudioTrackInfo(language: t.language, label: t.label, codecs: t.codecs, bitrate: t.bitrate, channelCount: t.channelCount, sampleRate: t.sampleRate, isMuxed: t.isMuxed, id: t.id, isSelected: index == idx)
        }
        let newSnapshot = TracksSnapshot(variants: snapshot.variants, audioTracks: updatedAudio, subtitleTracks: snapshot.subtitleTracks)
        sink.updateTrackList(info: newSnapshot)
        currentSnapshot = newSnapshot
        let activeAudio = updatedAudio.first { $0.isSelected }
        sink.updateActiveAudioTrack(info: activeAudio)
        hasSelectedAudio = activeAudio != nil
    }

    private func onNewErrorLogEntry() {
        guard let events = player?.currentItem?.errorLog()?.events else { return }
        while processedErrorEntries < events.count {
            let event = events[processedErrorEntries]
            sink.addPlaybackError(event: AVAccessLogMappersKt.loadError(
                nowMs: nowMs(),
                errorDomain: event.errorDomain,
                statusCode: Int64(event.errorStatusCode),
                uri: event.uri,
                comment: event.errorComment
            ))
            processedErrorEntries += 1
        }
    }

    // Monotonic wall-clock: CACurrentMediaTime() never goes backward; epochOffsetMs bridges it to
    // epoch time (set once per attach so NTP adjustments don't affect intra-session timestamps).
    private func nowMs() -> Int64 { Int64(CACurrentMediaTime() * 1000) + epochOffsetMs }
}
